Language/Haskell/GhcMod/FileMapping.hs

```diff
@@ -0,0 +1,67 @@
+module Language.Haskell.GhcMod.FileMapping
+    ( loadMappedFile
+    , loadMappedFiles
+    , unloadMappedFile
+    , mapFile
+    ) where
+
+import Language.Haskell.GhcMod.Types
+import Language.Haskell.GhcMod.Monad.Types
+import Language.Haskell.GhcMod.Gap
+import Language.Haskell.GhcMod.HomeModuleGraph
+
+import System.Directory
+import System.FilePath
+
+import Data.Time
+
+import Control.Monad.Trans.Maybe
+import GHC
+
+loadMappedFiles :: IOish m => GhcModT m ()
+loadMappedFiles = do
+  Options {fileMappings} <- options
+  mapM_ (uncurry loadMappedFile) $ reverse fileMappings
+
+loadMappedFile :: IOish m => FilePath -> FileMapping -> GhcModT m ()
+loadMappedFile from fm@(RedirectedMapping _) =
+  addToState from fm
+loadMappedFile from (MemoryMapping _) = do
+  let loop' acc = do
+        line <- getLine
```

Comment:
This seems the wrong place to do that piece of IO. Should really be in the frontend code.